# IMPORT NECESSARY PACKAGES

In [1]:
import numpy as np
import pandas as pd

## Creating identity square matrix (3 x 3)

In [2]:
I = np.identity(3)
I

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

## Reading a csv file from the given url
### copy the url of the csv file into a variable called url, which is the file path

In [3]:
url = "https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv"
fuel_data = pd.read_csv(url)
fuel_data.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [4]:
fuel_data.describe()

utility_id_ferc1   report_year  fuel_qty_burned  fuel_mmbtu_per_unit  \
count      29523.000000  29523.000000     2.952300e+04         29523.000000   
mean         118.601836   2005.806050     2.622119e+06             8.492111   
std           74.178353      7.025483     9.118004e+06            10.600220   
min            1.000000   1994.000000     1.000000e+00             0.000001   
25%           55.000000   2000.000000     1.381700e+04             1.024000   
50%          122.000000   2006.000000     2.533220e+05             5.762694   
75%          176.000000   2012.000000     1.424034e+06            17.006000   
max          514.000000   2018.000000     5.558942e+08           341.260000   

       fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
count               29523.000000                  2.952300e+04   
mean                  208.649031                  9.175704e+02   
std                  2854.490090                  6.877593e+04   
min                  -276.080000                 -8.749370e+02   
25%                     5.207000                  3.778500e+00   
50%                    26.000000                  1.737100e+01   
75%                    47.113000                  4.213700e+01   
max                139358.000000                  7.964521e+06   

       fuel_cost_per_mmbtu  
count         29523.000000  
mean             19.304354  
std            2091.540939  
min             -41.501000  
25%               1.940000  
50%               4.127000  
75%               7.745000  
max          359278.000000

## To get the category of the feature with the missing values and the imputation technique to be used

In [5]:
fuel_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
record_id                       29523 non-null object
utility_id_ferc1                29523 non-null int64
report_year                     29523 non-null int64
plant_name_ferc1                29523 non-null object
fuel_type_code_pudl             29523 non-null object
fuel_unit                       29343 non-null object
fuel_qty_burned                 29523 non-null float64
fuel_mmbtu_per_unit             29523 non-null float64
fuel_cost_per_unit_burned       29523 non-null float64
fuel_cost_per_unit_delivered    29523 non-null float64
fuel_cost_per_mmbtu             29523 non-null float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB


## Or

In [6]:
fuel_data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [7]:
fuel_data['fuel_unit'].dtypes

dtype('O')

## The above shows that the data type in fuel unit is object

## To get the second and third lowest correlation with the fuel Fuel Cost Per Unit Burned

In [8]:
fuel_data.corr()

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

## From the Above, the 'fuel qty burned' and 'fuel mmbtu per unit' has the second and third lowest correlation with the 'fuel cost per unit burned' respectively

In [9]:
fuel_data_1994 = fuel_data[fuel_data['report_year']==1994]
fuel_data_1994

record_id  utility_id_ferc1  report_year  \
0       f1_fuel_1994_12_1_0_7                 1         1994   
1      f1_fuel_1994_12_1_0_10                 1         1994   
2       f1_fuel_1994_12_2_0_1                 2         1994   
3       f1_fuel_1994_12_2_0_7                 2         1994   
4      f1_fuel_1994_12_2_0_10                 2         1994   
...                       ...               ...          ...   
1230  f1_fuel_1994_12_45_1_10                45         1994   
1231  f1_fuel_1994_12_45_1_13                45         1994   
1232   f1_fuel_1994_12_45_2_1                45         1994   
1233   f1_fuel_1994_12_45_2_4                45         1994   
1234   f1_fuel_1994_12_45_2_7                45         1994   

          plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0                 rockport                coal       ton        5377489.0   
1     rockport total plant                coal       ton       10486945.0   
2                   gorgas                coal       ton        2978683.0   
3                    barry                coal       ton        3739484.0   
4                chickasaw                 gas       mcf          40533.0   
...                    ...                 ...       ...              ...   
1230             riverbend                coal       ton         396121.0   
1231             riverbend                 gas       mcf          56258.0   
1232                  buck                coal       ton         205382.0   
1233                  buck                 gas       mcf          58828.0   
1234            buzz.roost                 gas       mcf          92365.0   

      fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0                  16.590                     18.590   
1                  16.592                     18.580   
2                  24.130                     39.720   
3                  23.950                     47.210   
4                   1.000                      2.770   
...                   ...                        ...   
1230               24.952                     39.950   
1231                1.036                      3.911   
1232               24.706                     39.190   
1233                1.034                      3.462   
1234                1.035                      3.790   

      fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                           18.530                1.121  
1                           18.530                1.120  
2                           38.120                1.650  
3                           45.990                1.970  
4                            2.770                2.570  
...                            ...                  ...  
1230                        43.700                1.601  
1231                         3.911                3.796  
1232                        39.370                1.586  
1233                         3.462                3.421  
1234                         3.790                3.685  

[1235 rows x 11 columns]

In [10]:
sum_fuel_cost = fuel_data_1994['fuel_cost_per_unit_burned'].sum()
sum_fuel_cost

263292.11500000005

In [11]:
fuel_data_1998 = fuel_data[fuel_data['report_year']==1998]
fuel_data_1998

record_id  utility_id_ferc1  report_year  \
4618     f1_fuel_1998_12_1_0_7                 1         1998   
4619    f1_fuel_1998_12_1_0_10                 1         1998   
4620     f1_fuel_1998_12_4_0_1                 4         1998   
4621     f1_fuel_1998_12_4_0_2                 4         1998   
4622     f1_fuel_1998_12_4_0_4                 4         1998   
...                        ...               ...          ...   
5720   f1_fuel_1998_12_145_2_2               145         1998   
5721   f1_fuel_1998_12_145_2_3               145         1998   
5722   f1_fuel_1998_12_145_2_6               145         1998   
5723   f1_fuel_1998_12_145_2_9               145         1998   
5724  f1_fuel_1998_12_145_2_12               145         1998   

          plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
4618    rockport total aeg                coal       ton        4965881.0   
4619  rockport total plant                coal       ton       10000938.0   
4620           units 1 - 3                coal       ton        1974682.0   
4621           units 1 - 3                 gas       mcf          43048.0   
4622                unit 4                coal       ton         385512.0   
...                    ...                 ...       ...              ...   
5720           fort lupton                 oil       bbl           3206.0   
5721           fort lupton                 gas       mcf         439412.0   
5722                fruita                 gas       mcf          52091.0   
5723             valmont 6                 gas       mcf         347619.0   
5724        fort st. vrain                 gas       mcf        5034730.0   

      fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
4618            17.198000                     19.282   
4619            17.196000                     19.280   
4620            22.266000                     26.550   
4621             1.031000                      2.285   
4622            22.330000                     26.550   
...                   ...                        ...   
5720             5.427366                     33.180   
5721             1.061000                      2.509   
5722             0.991000                      2.470   
5723             0.987000                      2.413   
5724             1.044000                      2.384   

      fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
4618                        19.338                1.126  
4619                        19.340                1.122  
4620                        26.540                1.190  
4621                         2.285                2.216  
4622                        26.540                1.190  
...                            ...                  ...  
5720                         0.000                6.113  
5721                         0.000                2.364  
5722                         0.000                2.493  
5723                         0.000                2.445  
5724                         0.000                2.282  

[1107 rows x 11 columns]

In [12]:
sum_fuel_cost_2 = fuel_data_1998['fuel_cost_per_unit_burned'].sum()
sum_fuel_cost_2

140238.65600000002

In [13]:
per_change = (sum_fuel_cost - sum_fuel_cost_2)/sum_fuel_cost*100
round(per_change,2)

46.74

## I think there is no answer  in the option for number 4.

# To know the year with the highest average cost, I will use groupby function

In [14]:
fuel_data.groupby('report_year')['fuel_cost_per_unit_delivered'].mean()

report_year
1994       63.636060
1995       32.735269
1996     9196.705948
1997    11140.197239
1998      287.154420
1999       25.551627
2000      985.362877
2001       60.050396
2002       47.594361
2003       55.663493
2004      139.524275
2005       41.438184
2006       38.657484
2007       43.325023
2008       58.588197
2009      652.694163
2010       91.862105
2011       59.774667
2012       60.994502
2013      172.307591
2014      192.737183
2015      326.535511
2016      103.901761
2017       46.196861
2018      499.269966
Name: fuel_cost_per_unit_delivered, dtype: float64

## From the above result, is can be seen that 1997 has the highest average

In [15]:
round(fuel_data['fuel_mmbtu_per_unit'].std(),2)

10.6

In [16]:
round(fuel_data['fuel_mmbtu_per_unit'].describe(),2)

count    29523.00
mean         8.49
std         10.60
min          0.00
25%          1.02
50%          5.76
75%         17.01
max        341.26
Name: fuel_mmbtu_per_unit, dtype: float64

## Therefore, the standard deviation is 10.60 and the 75th percentile is 17.01

In [17]:
round(fuel_data['fuel_qty_burned'].skew(),2)

15.85

In [18]:
round(fuel_data['fuel_qty_burned'].kurt(),2)

651.37

## Or

In [19]:
print('Skewness = ', round(fuel_data['fuel_qty_burned'].skew(), 2))
print('Kurtosis = ', round(fuel_data['fuel_qty_burned'].kurt(),2))

Skewness =  15.85
Kurtosis =  651.37


## The feature with the missing value, the total number of missing  value and the percentage of the missing row as a factor of the total number of rows in three decimal places

In [20]:
fuel_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
record_id                       29523 non-null object
utility_id_ferc1                29523 non-null int64
report_year                     29523 non-null int64
plant_name_ferc1                29523 non-null object
fuel_type_code_pudl             29523 non-null object
fuel_unit                       29343 non-null object
fuel_qty_burned                 29523 non-null float64
fuel_mmbtu_per_unit             29523 non-null float64
fuel_cost_per_unit_burned       29523 non-null float64
fuel_cost_per_unit_delivered    29523 non-null float64
fuel_cost_per_mmbtu             29523 non-null float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB


In [21]:
fuel_data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

## The feature that has the missing value = fuel unit, 
## The number of missing values = 180

In [22]:
n = fuel_data["fuel_unit"].isnull().sum()
n

180

In [23]:
d = fuel_data['report_year'].count()
d

29523

In [24]:
percentage = n*100/d
round(percentage,3)

0.61

## The percentage of  the missing row as a factor of the total number of rows in three decimal places is = 0.610 which is same as 0.61

In [25]:
fuel_data.groupby('fuel_type_code_pudl')["fuel_cost_per_unit_burned"].mean()

fuel_type_code_pudl
coal         67.421830
gas          13.659397
nuclear    4955.157002
oil         168.877086
other        18.253856
waste        19.518122
Name: fuel_cost_per_unit_burned, dtype: float64

## From the above result, it is seen that gas has the lowest average fuel cost per unit burned

## combinning lists using extend

In [26]:
A = [1,2,3,4,5,6] 
B = [13, 21, 34]
A.extend(B)
A

[1, 2, 3, 4, 5, 6, 13, 21, 34]

# THANKS A LOT FOR THE OPPORTUNITY TO IMPROVE UPTO  THIS LEVEL AND WILL BE EXPECTING MORE. 
# I BELIEVE MY SCORE IS 90% BECAUSE IT IS GLARING TO ME THAT I FAILED A QUESTION WHICH WAS DUE TO MISINTERPRETATION OF QUESTION.
# I  AM USING THIS AVENUE TO URGE HAMOYE DS INTERNSHIP TO PLEASE LOOK INTO IT CRITICALLY, BECAUSE I GOT 70% INSTEAD OF 80% OR 90%.
# KUDOS TO HAMOYE DATASCIENCE INTERNSHIP.